volatility3/framework/plugins/linux/kmsg.py

```diff
@@ -0,0 +1,387 @@
+        offset = text_data_ring.data + text_start
+
+        # Safety first ;)
+        text_len = min(info.text_len, end - begin)
+
+        return self.get_string(offset, text_len)
+
+    def get_log_lines(self, text_data_ring, desc, info) -> Generator[str, None, None]:
+        text = self.get_text_from_data_ring(text_data_ring, desc, info)
+        yield from text.splitlines()
+
+    def get_dict_lines(self, info) -> Generator[str, None, None]:
+        dict_text = utility.array_to_string(info.dev_info.subsystem)
+        if dict_text:
+            yield f" SUBSYSTEM={dict_text}"
+
+        dict_text = utility.array_to_string(info.dev_info.device)
+        if dict_text:
+            yield f" DEVICE={dict_text}"
+
+    def run(self) -> Iterator[Tuple[str, str, str, str, str]]:
+        # static struct printk_ringbuffer *prb = &printk_rb_static;
+        ringbuffers = self.vmlinux.object_from_symbol(symbol_name='prb').dereference()
+
+        desc_ring = ringbuffers.desc_ring
+        text_data_ring = ringbuffers.text_data_ring
+
+        desc_count = 1 << desc_ring.count_bits
+        desc_arr = self.vmlinux.object(object_type="array",
+                                       offset=desc_ring.descs,
+                                       subtype=self.vmlinux.get_type("prb_desc"),
+                                       count=desc_count)
+        info_arr = self.vmlinux.object(object_type="array",
+                                       offset=desc_ring.infos,
+                                       subtype=self.vmlinux.get_type("printk_info"),
+                                       count=desc_count)
+
+        # See kernel/printk/printk_ringbuffer.h
+        desc_state_var_bytes_sz = self.long_unsigned_int_size
+        desc_state_var_bits_sz = desc_state_var_bytes_sz * 8
+        desc_flags_shift = desc_state_var_bits_sz - 2
+        desc_flags_mask = 3 << desc_flags_shift
+        desc_id_mask = ~desc_flags_mask
+
+        cur_id = desc_ring.tail_id.counter
+        end_id = desc_ring.head_id.counter
+        while True:
+            desc = desc_arr[cur_id % desc_count]  # type: ignore
+            info = info_arr[cur_id % desc_count]  # type: ignore
+            desc_state = DescStateEnum((desc.state_var.counter >> desc_flags_shift) & 3)
+            if desc_state in (DescStateEnum.desc_committed, DescStateEnum.desc_finalized):
+                facility, level, timestamp, caller = self.get_prefix(info)
+                level_txt = self.get_level_text(level)
+                facility_txt = self.get_facility_text(facility)
+
+                for line in self.get_log_lines(text_data_ring, desc, info):
+                    yield facility_txt, level_txt, timestamp, caller, line
+                for line in self.get_dict_lines(info):
+                    yield facility_txt, level_txt, timestamp, caller, line
+
+            cur_id += 1
+            cur_id &= desc_id_mask
+            if cur_id == end_id:
```

Comment:
Since this comes at the end, I'd much prefer it if the `while True:` became `which cur_id != end_id:`.  I don't like the possibility of run away loops so `while True1 just always gives me the creeps...  5:)

Reply:
It's that Python needs a do/while ;)  lol ... hope now it looks better for you.